### ISSUE TITLE
[FR] Type hint implemention extends or interface inspection

### ISSUE THREAD

**Issue by Haehnchen** [kalessil/phpinspectionsea#265]
Some common pitfalls often seen: Instead of using interface implementations or abstract classes main implementation is used for type hints.

So collect all method calls and try to find a more suitable type hint. valuable feature but heavy code analysis ;)

```php
class Foobar() {
  public function __construct(Router $router) {
     $this->router = $router;
  }
  private foobar() {
    $this->router->generate();
  }
}
```

One common use case:

```php
class Router implements RouterInterface{
  public generate() {}
  public generateFoobar() {}
}
```

```php
class RouterInterface{
  public generate() {}
}
```

EDIT by VR: report final classes used in this context.

**Comment by funivan** [comment]
Implementation is not simple and it can generate a lot of false positive

**Comment by kalessil** [comment]
@funivan : complex, but we can just fix limitations, e.g: ensure the parameter is only consumed (methods calls, invoke context, string context). If the parameter is re-dispatched, we stopping the analysis.

**Comment by kalessil** [comment]
@Haehnchen : I'll try to pick up this feature in one of nearest release.

We are focusing on QFes and stabilization Q2/Q3, so depending on overall bug0reporting rate the feature can be shifted.

**Comment by kalessil** [comment]
Finally starting with this one...

**Comment by kalessil** [comment]
Reopening

**Comment by kalessil** [comment]
Done, currently, it's quite straight-forward and we'll need to find a good guessing criteria.
